### ISSUE TITLE
Have knife client bulk delete support 10k+ clients

### ISSUE THREAD

**Issue by mskdenigma** [chef/chef#10366]
### Describe the Enhancement:
Update `knife client bulk delete` or the Chef Server API to support Chef servers with 10k+ clients
Add the boolean telling you if a client is a validator or not to the client list API call
`GET /organizations/NAME/clients`

FROM:
```json
{
  "org1-validator" : "https://chef.example/orgaizations/org1/clients/org1-validator",
  "client1" : "https://chef.example/orgaizations/org1/clients/client1"
}
```

TO:
```json
{
  "org1-validator" : {
     "url" : "https://chef.example/orgaizations/org1/clients/org1-validator",
     "validator": true
   },
  "client1" : {
     "url" :  "https://chef.example/orgaizations/org1/clients/client1",
     "validator" : false
}
```

### Describe the Need:
Currently [client_bulk_delete.rb](https://github.com/chef/chef/blob/master/lib/chef/knife/client_bulk_delete.rb) does the following to achieve its task of bulk deleting clients

1. Get all the clients from the chef server (Currently accomplished as doing a search for all clients matching the pattern *:*)
2. Loop through this hash of client objects and confirm if they match the regex requested
3.  After confirming the match sort clients into arrays of normal clients or validator clients and act upon them according to whether or not you set the -D flag

The issue is in step 1, when you have 10k+ clients you end up trying to make a search that is way too big for elasticsearch, at least in its default settings. This actually is caused by calling the the [client list function](https://github.com/chef/chef/blob/1375318db8d92f88f4c5814bb31b9252d445bc7c/lib/chef/api_client_v1.rb#L191) with the  inflate boolean set to [true](https://github.com/chef/chef/blob/1375318db8d92f88f4c5814bb31b9252d445bc7c/lib/chef/knife/client_bulk_delete.rb#L41).

Now the reason this is done as I found out is because when you search for the client you get a client object that is a hash that includes information such as boolean indicating if it is a validator or not. If inflate is set to false you make make an api call to just get the entire list of clients which is very fast and less taxing on the chef server/elasticsearch backend, BUT does not tell you if the client is a validator. So there are downsides with both approaches

- Getting Client List via Search : searching for the list of clients is very taxing, and near impossible for 10k+ clients, in fact will not work at all with default elasticsearch settings
- Getting Client List via `GET /organizations/NAME/clients` : very fast and less taxing BUT will require you to then run an API call for EVERY client after matching regex to determine if they are a validator

### Current Alternative
Alternative solutions to this include running a patched version of the chef libraries where I set `inflate` to false which runs the get client list API and then patching [client_bulk_delete.rb](https://github.com/chef/chef/blob/master/lib/chef/knife/client_bulk_delete.rb) to run a get client API call for each client that matched the regex, this doubles how long it takes to bulk delete clients BUT is significantly less strenuous to the server.

Another solution is having the chef server check if the client is a validator when receiving a DELETE request and requiring a boolean with the request that is only set to true when the user passes the -D flag to the CLI, something like

`DELETE /organizations/NAME/clients/NAME`
```json
{
    "delete_validator" : true
}
```

### Can We Help You Implement This?:
I'd like to get input as to whether the proposed change of getting the validator boolean with `GET /organizations/NAME/clients` makes sense or if it is like a security red flag. Otherwise I will make PR for the solution in Current Alternative.

**Comment by lamont-granquist** [comment]
Feel free to fix this, it isn't a security issue.  It was done because it was somewhat difficult to figure out how to do more than 10k nodes, so it was just capped and the issue kicked down the road.  I think we also expected anyone who had that many nodes to just write a SQL query directly against the database to clean it up.

The idea of doing an API request per node though in bulk delete is probably not what anyone expects in bulk delete though.  I would expect the search to use pagination and partial search results.  If that blows up elasticsearch with the default settings then something should be fixed there with the settings or the implementation in the server.

Oh you super-duper cannot change the client list results from a String to a Hash.  That's a breaking change and not compelling enough.

**Comment by tas50** [comment]
Thanks for submitting this issue. After review by the team, we've decided that this isn't something that we're going to work on ourselves. We receive dozens of new issues a month, which is more than our team can possibly work on. This means we must prioritize work and be honest about what we're able to get done as a team.  That doesn't mean that we think this is a bad idea or isn't a necessary change. It's just not something we'll likely get to given competing priorities. I've applied our `Design Proposal: Accepted. PRs Welcome` label to this issue and closed it out. If you or someone else would like to work on this issue we're open to pull requests. If you'd like to chat further about what that design might look reach out to us in our [#chef-infra-dev](https://chefcommunity.slack.com/archives/C55FCUGBS) slack channel.
